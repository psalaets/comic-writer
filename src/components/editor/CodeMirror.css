/******************************************************************************/
/* CodeMirror                                                                 */
/******************************************************************************/
/*
 * 1. The superflurous attribute selector is here to out-specify a codemirror
 *    selector.The codemirror selector is `.CodeMirror-wrap pre` (011) and this
 *    is 020
 * 2. the width in "X's" on each side.
 * 3. Each Line of code.
 * 4. Wraps the text in each line's gutter value.
 * 5. Turning off "focus" outline for the editor.
 *    TODO: What's the right way to make this a11y without it looking like total
 *          shit?
 * 6. Resetting z-index to creat a new stacking context for this Component.
*/

/* Block */
/******************************************************************************/
.c-codemirror {
  background-color: var(--color-utility-pure);
  border-left: 1px solid var(--color-utility-milk);
  border-right: 1px solid var(--color-utility-milk);
  height: 100%;
  margin-left: auto;
  margin-right: auto;
  padding-left: 1.5ex; /* 2 */
  padding-right: 3ex; /* 2 */
  width: 100%;
  z-index: 0; /* 6 */
}

/* Misc */
/******************************************************************************/
.CodeMirror-placeholder[class] { /* 1 */
  color: var(--color-utility-dove);
}

.CodeMirror-line[class] {} /* 3 */

.CodeMirror-gutter-elt { /* 4 */
  align-items: flex-start;
  bottom: 0;
  display: flex;
  top: 0;
}

.CodeMirror-gutter-wrapper {
  bottom: 0;
  top: 0;
}

.CodeMirror *:focus { /* 5 */
  outline: none;
}

/* panel count line widget */
.panel-count {
  color: var(--color-utility-dove);
}
