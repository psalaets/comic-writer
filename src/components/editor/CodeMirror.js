import React, { Component } from 'react';
import PropTypes from 'prop-types';

import codemirror from 'codemirror';
import 'codemirror/lib/codemirror.css';
import './CodeMirror.css';

import { MODE, THEME } from '../../codemirror/comic-writer-mode';
import {
  ID as WORD_COUNTS,
  create as createWordCounts
} from '../../codemirror/gutters/word-counts';

import {
  create as createPanelCounts
} from '../../codemirror/panel-counts';


import '../../codemirror/theme.css';

import 'codemirror/addon/display/placeholder';
import 'codemirror/addon/scroll/scrollpastend';

export default class CodeMirror extends Component {
  constructor(props) {
    super(props);

    this.el = React.createRef();
  }

  render() {
    const styles = {
      maxWidth: `${this.props.editorWidth + 2}ex`,
    }
    return <div className="c-codemirror" style={styles} ref={this.el} />;
  }

  componentDidUpdate(prevProps) {
    const valueChanged = prevProps.value !== this.props.value;
    const statsChanged = prevProps.stats !== this.props.stats;

    if (valueChanged) {
      if (!this.cm.getValue()) {
        this.cm.setValue(this.props.value);
      } else {
        const oldLines = this.cm.getValue().split('\n');
        const newLines = this.props.value.split('\n');

        this.cm.operation(() => {
          newLines.forEach((newLine, index) => {
            const oldLine = oldLines[index] || '';
            if (newLine !== oldLine) {
              const from = { line: index, ch: 0 };
              const to = { line: index, ch: 10000 };

              this.cm.replaceRange(newLine, from, to, 'setValue');
            }
          });
        });
      }
    }

    if (statsChanged) {
      this.wordCountsGutter.update(this.props.stats);
      this.panelCounts.update(this.props.stats);
    }
  }

  componentDidMount() {
    this.cm = codemirror(this.el.current, {
      mode: MODE,
      theme: THEME,
      value: this.props.value,
      inputStyle: 'contenteditable',
      placeholder: 'Adventure starts here...',
      lineWrapping: true,
      cursorScrollMargin: 100, // Not *exactly* sure why this value works.
      scrollbarStyle: null,
      scrollPastEnd: true,
      gutters: [WORD_COUNTS]
    });

    this.cm.setSize('100%', '100%');

    this.cm.on('change', (cm, change) => {
      if (change.origin === 'setValue') {
        return;
      }

      this.props.onChange({
        value: cm.getValue()
      })
    });

    this.wordCountsGutter = createWordCounts(this.cm);
    this.panelCounts = createPanelCounts(this.cm);
  }

  componentWillUnmount() {

  }
}

CodeMirror.propTypes = {
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  stats: PropTypes.array.isRequired
};
