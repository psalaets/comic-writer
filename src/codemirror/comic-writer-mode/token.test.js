import CodeMirror from 'codemirror';
import createMode from './mode';

const cmConfig = {};
const mode = createMode(cmConfig);

const startState = mode.startState;
const token = mode.token;

describe('mode.token()', () => {
  describe('page', () => {
    test('single digit page', () => {
      const tokens = collectTokens('page 5');

      expect(tokens).toMatchSnapshot();
    });

    test('multi digit page', () => {
      const tokens = collectTokens('page 10');

      expect(tokens).toMatchSnapshot();
    });

    test('without suffix', () => {
      const tokens = collectTokens('page');

      expect(tokens).toMatchSnapshot();
    });

    test('without suffix number', () => {
      const tokens = collectTokens('page ');

      expect(tokens).toMatchSnapshot();
    });
  });

  describe('panel', () => {
    test('single digit panel', () => {
      const tokens = collectTokens('panel 2');

      expect(tokens).toMatchSnapshot();
    });

    test('multi digit panel', () => {
      const tokens = collectTokens('panel 10');

      expect(tokens).toMatchSnapshot();
    });

    test('without suffix', () => {
      const tokens = collectTokens('panel');

      expect(tokens).toMatchSnapshot();
    });

    test('without suffix number', () => {
      const tokens = collectTokens('panel ');

      expect(tokens).toMatchSnapshot();
    });
  });

  describe('sfx', () => {
    test('full', () => {
      const tokens = collectTokens('\tsfx (gun): blam');

      expect(tokens).toMatchSnapshot();
    });

    test('no modifier', () => {
      const tokens = collectTokens('\tsfx: blam');

      expect(tokens).toMatchSnapshot();
    });

    test('newly opened modifier', () => {
      const tokens = collectTokens('\tsfx (');

      expect(tokens).toMatchSnapshot();
    });

    test('unclosed modifier', () => {
      const tokens = collectTokens('\tsfx (gun');

      expect(tokens).toMatchSnapshot();
    });

    test('closed modifier but no colon', () => {
      const tokens = collectTokens('\tsfx (gun)');

      expect(tokens).toMatchSnapshot();
    });

    test('no content', () => {
      const tokens = collectTokens('\tsfx:');

      expect(tokens).toMatchSnapshot();
    });

    test('no colon', () => {
      const tokens = collectTokens('\tsfx');

      expect(tokens).toMatchSnapshot();
    });

    test('allows bold content', () => {
      const tokens = collectTokens('\tsfx: bam **bam** bam');

      expect(tokens).toMatchSnapshot();
    });
  });

  describe('caption', () => {
    test('full', () => {
      const tokens = collectTokens('\tcaption (bill): it went well');

      expect(tokens).toMatchSnapshot();
    });

    test('no modifier', () => {
      const tokens = collectTokens('\tcaption: it went well');

      expect(tokens).toMatchSnapshot();
    });

    test('all bold content', () => {
      const tokens = collectTokens('\tcaption: **good one**');

      expect(tokens).toMatchSnapshot();
    });

    test('some bold content', () => {
      const tokens = collectTokens('\tcaption: it **did not** work');

      expect(tokens).toMatchSnapshot();
    });

    test('multiple bolds', () => {
      const tokens = collectTokens('\tcaption: it did **not** work at **all**!');

      expect(tokens).toMatchSnapshot();
    });

    test('no content', () => {
      const tokens = collectTokens('\tcaption:');

      expect(tokens).toMatchSnapshot();
    });

    test('no colon', () => {
      const tokens = collectTokens('\tcaption');

      expect(tokens).toMatchSnapshot();
    });

    test('newly opened modifier', () => {
      const tokens = collectTokens('\tcaption (');

      expect(tokens).toMatchSnapshot();
    });

    test('unclosed modifier', () => {
      const tokens = collectTokens('\tcaption (mom');

      expect(tokens).toMatchSnapshot();
    });

    test('closed modifier but no colon', () => {
      const tokens = collectTokens('\tcaption (mom)');

      expect(tokens).toMatchSnapshot();
    });
  });

  describe('dialogue', () => {
    test('full', () => {
      const tokens = collectTokens('\tbill (whisper): it went well');

      expect(tokens).toMatchSnapshot();
    });

    test('no modifier', () => {
      const tokens = collectTokens('\tbill: it went well');

      expect(tokens).toMatchSnapshot();
    });

    test('all bold content', () => {
      const tokens = collectTokens('\tbill: **good one**');

      expect(tokens).toMatchSnapshot();
    });

    test('some bold content', () => {
      const tokens = collectTokens('\tbill: it **did not** work');

      expect(tokens).toMatchSnapshot();
    });

    test('multiple bolds', () => {
      const tokens = collectTokens('\tbill: it did **not** work at **all**!');

      expect(tokens).toMatchSnapshot();
    });

    test('no content', () => {
      const tokens = collectTokens('\tbill:');

      expect(tokens).toMatchSnapshot();
    });

    test('no colon', () => {
      const tokens = collectTokens('\tbill');

      expect(tokens).toMatchSnapshot();
    });

    test('newly opened modifier', () => {
      const tokens = collectTokens('\tbill (');

      expect(tokens).toMatchSnapshot();
    });

    test('unclosed modifier', () => {
      const tokens = collectTokens('\tbill (yell');

      expect(tokens).toMatchSnapshot();
    });

    test('unclosed modifier with content', () => {
      const tokens = collectTokens('\tbill (yell: ahhh');

      expect(tokens).toMatchSnapshot();
    });

    test('unopened modifier', () => {
      const tokens = collectTokens('\tbill yell)');

      expect(tokens).toMatchSnapshot();
    });

    test('unopened modifier with content', () => {
      const tokens = collectTokens('\tbill yell): ahhh');

      expect(tokens).toMatchSnapshot();
    });

    test('closed modifier but no colon', () => {
      const tokens = collectTokens('\tbill (yell)');

      expect(tokens).toMatchSnapshot();
    });

    test('no subject', () => {
      const tokens = collectTokens('\t(yell): bam');

      expect(tokens).toMatchSnapshot();
    });
  });

  describe('metadata', () => {
    test('full', () => {
      const tokens = collectTokens('name: value');

      expect(tokens).toMatchSnapshot();
    });

    test('no space after colon', () => {
      const tokens = collectTokens('name:value');

      expect(tokens).toMatchSnapshot();
    });

    test('no value', () => {
      const tokens = collectTokens('name:');

      expect(tokens).toMatchSnapshot();
    });

    test('no colon', () => {
      const tokens = collectTokens('name');

      expect(tokens).toMatchSnapshot();
    });
  });

  describe('paragraph', () => {
    test('text', () => {
      const tokens = collectTokens('The first rays of sunrise hit the house. A car waits out front.');

      expect(tokens).toMatchSnapshot();
    });
  });
});

function collectTokens(line) {
  const tabSize = 4;
  const stream = new CodeMirror.StringStream(line, tabSize);
  const state = startState();

  const tokens = [];

  while (!stream.eol()) {
    stream.start = stream.pos;
    const style = token(stream, state);

    if (stream.pos === stream.start) {
      throw Error('failed to advance stream');
    }

    tokens.push({
      style: style ? style.split(' ').sort() : style,
      start: stream.start,
      end: stream.pos,
      string: stream.current()
    });
  }

  return tokens;
}