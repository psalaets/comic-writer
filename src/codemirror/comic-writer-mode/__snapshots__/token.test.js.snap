// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mode.token() caption all bold content 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-caption",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 8,
    "start": 1,
    "string": "caption",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 9,
    "start": 8,
    "string": ":",
    "style": null,
  },
  Object {
    "end": 10,
    "start": 9,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 22,
    "start": 10,
    "string": "**good one**",
    "style": Array [
      "lettering-bold",
      "lettering-content",
    ],
  },
]
`;

exports[`mode.token() caption closed modifier but no colon 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-caption",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 8,
    "start": 1,
    "string": "caption",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 9,
    "start": 8,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 14,
    "start": 9,
    "string": "(mom)",
    "style": Array [
      "lettering-modifier",
    ],
  },
]
`;

exports[`mode.token() caption full 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-caption",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 8,
    "start": 1,
    "string": "caption",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 9,
    "start": 8,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 15,
    "start": 9,
    "string": "(bill)",
    "style": Array [
      "lettering-modifier",
    ],
  },
  Object {
    "end": 16,
    "start": 15,
    "string": ":",
    "style": null,
  },
  Object {
    "end": 17,
    "start": 16,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 29,
    "start": 17,
    "string": "it went well",
    "style": Array [
      "lettering-content",
    ],
  },
]
`;

exports[`mode.token() caption multiple bolds 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-caption",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 8,
    "start": 1,
    "string": "caption",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 9,
    "start": 8,
    "string": ":",
    "style": null,
  },
  Object {
    "end": 10,
    "start": 9,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 17,
    "start": 10,
    "string": "it did ",
    "style": Array [
      "lettering-content",
    ],
  },
  Object {
    "end": 24,
    "start": 17,
    "string": "**not**",
    "style": Array [
      "lettering-bold",
      "lettering-content",
    ],
  },
  Object {
    "end": 33,
    "start": 24,
    "string": " work at ",
    "style": Array [
      "lettering-content",
    ],
  },
  Object {
    "end": 40,
    "start": 33,
    "string": "**all**",
    "style": Array [
      "lettering-bold",
      "lettering-content",
    ],
  },
  Object {
    "end": 41,
    "start": 40,
    "string": "!",
    "style": Array [
      "lettering-content",
    ],
  },
]
`;

exports[`mode.token() caption newly opened modifier 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-caption",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 8,
    "start": 1,
    "string": "caption",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 9,
    "start": 8,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 10,
    "start": 9,
    "string": "(",
    "style": Array [
      "lettering-modifier",
    ],
  },
]
`;

exports[`mode.token() caption no colon 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-caption",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 8,
    "start": 1,
    "string": "caption",
    "style": Array [
      "lettering-subject",
    ],
  },
]
`;

exports[`mode.token() caption no content 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-caption",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 8,
    "start": 1,
    "string": "caption",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 9,
    "start": 8,
    "string": ":",
    "style": null,
  },
]
`;

exports[`mode.token() caption no modifier 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-caption",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 8,
    "start": 1,
    "string": "caption",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 9,
    "start": 8,
    "string": ":",
    "style": null,
  },
  Object {
    "end": 10,
    "start": 9,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 22,
    "start": 10,
    "string": "it went well",
    "style": Array [
      "lettering-content",
    ],
  },
]
`;

exports[`mode.token() caption some bold content 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-caption",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 8,
    "start": 1,
    "string": "caption",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 9,
    "start": 8,
    "string": ":",
    "style": null,
  },
  Object {
    "end": 10,
    "start": 9,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 13,
    "start": 10,
    "string": "it ",
    "style": Array [
      "lettering-content",
    ],
  },
  Object {
    "end": 24,
    "start": 13,
    "string": "**did not**",
    "style": Array [
      "lettering-bold",
      "lettering-content",
    ],
  },
  Object {
    "end": 29,
    "start": 24,
    "string": " work",
    "style": Array [
      "lettering-content",
    ],
  },
]
`;

exports[`mode.token() caption unclosed modifier 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-caption",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 8,
    "start": 1,
    "string": "caption",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 9,
    "start": 8,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 13,
    "start": 9,
    "string": "(mom",
    "style": Array [
      "lettering-modifier",
    ],
  },
]
`;

exports[`mode.token() dialogue all bold content 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-dialogue",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 5,
    "start": 1,
    "string": "bill",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 6,
    "start": 5,
    "string": ":",
    "style": null,
  },
  Object {
    "end": 7,
    "start": 6,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 19,
    "start": 7,
    "string": "**good one**",
    "style": Array [
      "lettering-bold",
      "lettering-content",
    ],
  },
]
`;

exports[`mode.token() dialogue closed modifier but no colon 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-dialogue",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 5,
    "start": 1,
    "string": "bill",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 6,
    "start": 5,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 12,
    "start": 6,
    "string": "(yell)",
    "style": Array [
      "lettering-modifier",
    ],
  },
]
`;

exports[`mode.token() dialogue full 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-dialogue",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 5,
    "start": 1,
    "string": "bill",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 6,
    "start": 5,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 15,
    "start": 6,
    "string": "(whisper)",
    "style": Array [
      "lettering-modifier",
    ],
  },
  Object {
    "end": 16,
    "start": 15,
    "string": ":",
    "style": null,
  },
  Object {
    "end": 17,
    "start": 16,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 29,
    "start": 17,
    "string": "it went well",
    "style": Array [
      "lettering-content",
    ],
  },
]
`;

exports[`mode.token() dialogue multiple bolds 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-dialogue",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 5,
    "start": 1,
    "string": "bill",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 6,
    "start": 5,
    "string": ":",
    "style": null,
  },
  Object {
    "end": 7,
    "start": 6,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 14,
    "start": 7,
    "string": "it did ",
    "style": Array [
      "lettering-content",
    ],
  },
  Object {
    "end": 21,
    "start": 14,
    "string": "**not**",
    "style": Array [
      "lettering-bold",
      "lettering-content",
    ],
  },
  Object {
    "end": 30,
    "start": 21,
    "string": " work at ",
    "style": Array [
      "lettering-content",
    ],
  },
  Object {
    "end": 37,
    "start": 30,
    "string": "**all**",
    "style": Array [
      "lettering-bold",
      "lettering-content",
    ],
  },
  Object {
    "end": 38,
    "start": 37,
    "string": "!",
    "style": Array [
      "lettering-content",
    ],
  },
]
`;

exports[`mode.token() dialogue newly opened modifier 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-dialogue",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 5,
    "start": 1,
    "string": "bill",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 6,
    "start": 5,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 7,
    "start": 6,
    "string": "(",
    "style": Array [
      "lettering-modifier",
    ],
  },
]
`;

exports[`mode.token() dialogue no colon 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-dialogue",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 5,
    "start": 1,
    "string": "bill",
    "style": Array [
      "lettering-subject",
    ],
  },
]
`;

exports[`mode.token() dialogue no content 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-dialogue",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 5,
    "start": 1,
    "string": "bill",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 6,
    "start": 5,
    "string": ":",
    "style": null,
  },
]
`;

exports[`mode.token() dialogue no modifier 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-dialogue",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 5,
    "start": 1,
    "string": "bill",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 6,
    "start": 5,
    "string": ":",
    "style": null,
  },
  Object {
    "end": 7,
    "start": 6,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 19,
    "start": 7,
    "string": "it went well",
    "style": Array [
      "lettering-content",
    ],
  },
]
`;

exports[`mode.token() dialogue some bold content 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-dialogue",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 5,
    "start": 1,
    "string": "bill",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 6,
    "start": 5,
    "string": ":",
    "style": null,
  },
  Object {
    "end": 7,
    "start": 6,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 10,
    "start": 7,
    "string": "it ",
    "style": Array [
      "lettering-content",
    ],
  },
  Object {
    "end": 21,
    "start": 10,
    "string": "**did not**",
    "style": Array [
      "lettering-bold",
      "lettering-content",
    ],
  },
  Object {
    "end": 26,
    "start": 21,
    "string": " work",
    "style": Array [
      "lettering-content",
    ],
  },
]
`;

exports[`mode.token() dialogue unclosed modifier 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-dialogue",
      "line-cm-lettering",
    ],
  },
  Object {
    "end": 5,
    "start": 1,
    "string": "bill",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 6,
    "start": 5,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 11,
    "start": 6,
    "string": "(yell",
    "style": Array [
      "lettering-modifier",
    ],
  },
]
`;

exports[`mode.token() metadata full 1`] = `
Array [
  Object {
    "end": 11,
    "start": 0,
    "string": "name: value",
    "style": Array [
      "metadata",
    ],
  },
]
`;

exports[`mode.token() metadata no colon 1`] = `
Array [
  Object {
    "end": 4,
    "start": 0,
    "string": "name",
    "style": null,
  },
]
`;

exports[`mode.token() metadata no space after colon 1`] = `
Array [
  Object {
    "end": 10,
    "start": 0,
    "string": "name:value",
    "style": Array [
      "metadata",
    ],
  },
]
`;

exports[`mode.token() metadata no value 1`] = `
Array [
  Object {
    "end": 5,
    "start": 0,
    "string": "name:",
    "style": null,
  },
]
`;

exports[`mode.token() page multi digit page 1`] = `
Array [
  Object {
    "end": 7,
    "start": 0,
    "string": "page 10",
    "style": Array [
      "page",
    ],
  },
]
`;

exports[`mode.token() page single digit page 1`] = `
Array [
  Object {
    "end": 6,
    "start": 0,
    "string": "page 5",
    "style": Array [
      "page",
    ],
  },
]
`;

exports[`mode.token() page without suffix 1`] = `
Array [
  Object {
    "end": 4,
    "start": 0,
    "string": "page",
    "style": null,
  },
]
`;

exports[`mode.token() page without suffix number 1`] = `
Array [
  Object {
    "end": 5,
    "start": 0,
    "string": "page ",
    "style": null,
  },
]
`;

exports[`mode.token() panel multi digit panel 1`] = `
Array [
  Object {
    "end": 8,
    "start": 0,
    "string": "panel 10",
    "style": Array [
      "panel",
    ],
  },
]
`;

exports[`mode.token() panel single digit panel 1`] = `
Array [
  Object {
    "end": 7,
    "start": 0,
    "string": "panel 2",
    "style": Array [
      "panel",
    ],
  },
]
`;

exports[`mode.token() panel without suffix 1`] = `
Array [
  Object {
    "end": 5,
    "start": 0,
    "string": "panel",
    "style": null,
  },
]
`;

exports[`mode.token() panel without suffix number 1`] = `
Array [
  Object {
    "end": 6,
    "start": 0,
    "string": "panel ",
    "style": null,
  },
]
`;

exports[`mode.token() paragraph text 1`] = `
Array [
  Object {
    "end": 63,
    "start": 0,
    "string": "The first rays of sunrise hit the house. A car waits out front.",
    "style": null,
  },
]
`;

exports[`mode.token() sfx closed modifier but no colon 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-lettering",
      "line-cm-sfx",
    ],
  },
  Object {
    "end": 4,
    "start": 1,
    "string": "sfx",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 5,
    "start": 4,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 10,
    "start": 5,
    "string": "(gun)",
    "style": Array [
      "lettering-modifier",
    ],
  },
]
`;

exports[`mode.token() sfx does not allow bold content 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-lettering",
      "line-cm-sfx",
    ],
  },
  Object {
    "end": 4,
    "start": 1,
    "string": "sfx",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 5,
    "start": 4,
    "string": ":",
    "style": null,
  },
  Object {
    "end": 6,
    "start": 5,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 21,
    "start": 6,
    "string": "bam **bam** bam",
    "style": Array [
      "lettering-content",
    ],
  },
]
`;

exports[`mode.token() sfx full 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-lettering",
      "line-cm-sfx",
    ],
  },
  Object {
    "end": 4,
    "start": 1,
    "string": "sfx",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 5,
    "start": 4,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 10,
    "start": 5,
    "string": "(gun)",
    "style": Array [
      "lettering-modifier",
    ],
  },
  Object {
    "end": 11,
    "start": 10,
    "string": ":",
    "style": null,
  },
  Object {
    "end": 12,
    "start": 11,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 16,
    "start": 12,
    "string": "blam",
    "style": Array [
      "lettering-content",
    ],
  },
]
`;

exports[`mode.token() sfx newly opened modifier 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-lettering",
      "line-cm-sfx",
    ],
  },
  Object {
    "end": 4,
    "start": 1,
    "string": "sfx",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 5,
    "start": 4,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 6,
    "start": 5,
    "string": "(",
    "style": Array [
      "lettering-modifier",
    ],
  },
]
`;

exports[`mode.token() sfx no colon 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-lettering",
      "line-cm-sfx",
    ],
  },
  Object {
    "end": 4,
    "start": 1,
    "string": "sfx",
    "style": Array [
      "lettering-subject",
    ],
  },
]
`;

exports[`mode.token() sfx no content 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-lettering",
      "line-cm-sfx",
    ],
  },
  Object {
    "end": 4,
    "start": 1,
    "string": "sfx",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 5,
    "start": 4,
    "string": ":",
    "style": null,
  },
]
`;

exports[`mode.token() sfx no modifier 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-lettering",
      "line-cm-sfx",
    ],
  },
  Object {
    "end": 4,
    "start": 1,
    "string": "sfx",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 5,
    "start": 4,
    "string": ":",
    "style": null,
  },
  Object {
    "end": 6,
    "start": 5,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 10,
    "start": 6,
    "string": "blam",
    "style": Array [
      "lettering-content",
    ],
  },
]
`;

exports[`mode.token() sfx unclosed modifier 1`] = `
Array [
  Object {
    "end": 1,
    "start": 0,
    "string": "	",
    "style": Array [
      "line-cm-lettering",
      "line-cm-sfx",
    ],
  },
  Object {
    "end": 4,
    "start": 1,
    "string": "sfx",
    "style": Array [
      "lettering-subject",
    ],
  },
  Object {
    "end": 5,
    "start": 4,
    "string": " ",
    "style": null,
  },
  Object {
    "end": 9,
    "start": 5,
    "string": "(gun",
    "style": Array [
      "lettering-modifier",
    ],
  },
]
`;
